PQC: integrate ML-KEM and ML-DSA changes from draft-10 (past last call)

larabr opened the pull request:
https://datatracker.ietf.org/doc/draft-ietf-openpgp-pqc/10/

larabr committed:
PQC: allow v4 ML-KEM+X25119 encryption subkeys

larabr committed:
PQC: switch to non-experimental codepoints for ML-DSA and ML-KEM

larabr committed:
PQC: update KEM key combiner (draft-08)

larabr committed:
PQC: remove hash algo binding for ML-DSA signatures

larabr committed:
PQC: update test vectors to draft-09

twiss commented on src/key/helper.js:
```diff
@@ -117,12 +117,6 @@ export async function createBindingSignature(subkey, primaryKey, options, config
  * @async
  */
 export async function getPreferredHashAlgo(targetKeys, signingKeyPacket, date = new Date(), targetUserIDs = [], config) {
-  if (signingKeyPacket.algorithm === enums.publicKey.pqc_mldsa_ed25519) {
-    // For PQC, the returned hash algo MUST be set to the specified algorithm, see
-    // https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-pqc#section-5.2.1.
-    return publicKey.postQuantum.signature.getRequiredHashAlgo(signingKeyPacket.algorithm);
-  }
```
I know this is not required anymore but perhaps we should still do something like this, to match the security strength of the algorithm, similar to what we do for ECC below?

larabr replied:
Waiting for https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/188

larabr committed:
PQC signatures: select and require hash digest of at least 256 bits (draft-10)

See https://www.ietf.org/archive/id/draft-ietf-openpgp-pqc-10.html#section-9.4 .

When signing, the `preferredHashAlgo` is used as long as it's supported and long enough.
Otherwise, we look at the strongest supported algo,
and ultimately fallback the default algo (SHA-256).

twiss commented on src/crypto/public_key/elliptic/eddsa.js:
```diff
@@ -77,13 +77,15 @@ export async function generate(algo) {
  */
-export async function sign(algo, hashAlgo, message, publicKey, privateKey, hashed) {
-  if (getHashByteLength(hashAlgo) < getHashByteLength(getPreferredHashAlgo(algo))) {
+export async function sign(algo, hashAlgo, message, publicKey, privateKey, hashed, disableHashLengthCheck = false) {
+  if (!disableHashLengthCheck && getHashByteLength(hashAlgo) < getHashByteLength(getPreferredHashAlgo(algo))) {
```
Rather than disabling this check here, I think it'd be cleaner to move this check to the Ed25519 and Ed448 algorithms. That way, the code might also become simpler, e.g. for Ed25519 you could just do
```js
  if (getHashByteLength(hashAlgo) < 32) {
```
exactly as the spec says, rather than calling out to `getPreferredHashAlgo`.

larabr replied:
The code is currently shared between ed25519 and 448, and with the WebCrypto ed448 support in sight I think there is going to be almost 100% code overlap between them in the future; perhaps here we could take the min hash size as input (defaulting to `getHashByteLength(getPreferredHashAlgo(algo))` instead of the flag to disable the check.
So yeah let's delay this topic until upstream integration, also considering there is another upcoming PR there for WebCrypto integration that affects the eddsa code as well 🙂

twiss replied:
I didn't mean to suggest to duplicate this whole function, really only just the check for the hash function length :)

larabr replied:
Ah, but then the problem with the incompatible check would remain when we integrate PQC+Ed448 no?

twiss replied:
No, my idea was to move the check to the algorithms [here](https://github.com/openpgpjs/openpgpjs/blob/2a8969b4370a7f73792faa4ac67d2e4bc2ec474d/src/crypto/signature.js#L173) and [here](https://github.com/openpgpjs/openpgpjs/blob/2a8969b4370a7f73792faa4ac67d2e4bc2ec474d/src/crypto/signature.js#L119) (I see now that the code is shared also there so there's not even a need to duplicate it). That way, it doesn't affect PQC+EdDSA.

larabr replied:
Mm ok, I didn't want to distribute the checks around different modules, but I see we do something similar [here](https://github.com/openpgpjs/openpgpjs/blob/448418a6f56ece88dfb28c26a85107b887061d9d/src/crypto/crypto.js#L68-L71) so I can move this one too, since it's openpgp specific behavior anyway 👍

twiss commented on src/crypto/public_key/post_quantum/signature/signature.js:
```diff
@@ -52,14 +51,10 @@ export async function verify(signatureAlgo, hashAlgo, eccPublicKey, mldsaPublicK
-    default:
-      throw new Error('Unsupported signature algorithm');
-  }
+export function isCompatibleHashAlgo(signatureAlgo, hashAlgo) {
+  // The signature hash algo MUST be have digest larger than 256 bits
+  // https://www.ietf.org/archive/id/draft-ietf-openpgp-pqc-10.html#section-9.4
+  return getHashByteLength(hashAlgo) >= 32;
 }
```
Since the `signatureAlgo` is unused, I'd just remove it.

twiss commented on src/crypto/public_key/post_quantum/signature/signature.js:
```diff
@@ -15,12 +16,11 @@ export async function generate(algo) {
-    // https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-pqc#section-5.2.1.
-    throw new Error('Unexpected hash algorithm for PQC signature');
+  if (!isCompatibleHashAlgo(signatureAlgo, hashAlgo)) {
+    // The signature hash algo MUST be have digest larger than 256 bits
```
```suggestion
    // The signature hash algo MUST have digest larger than 256 bits
```

twiss commented on src/crypto/public_key/post_quantum/signature/signature.js:
```diff
@@ -34,12 +34,11 @@ export async function sign(signatureAlgo, hashAlgo, eccSecretKey, eccPublicKey,
-    // https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-pqc#section-5.2.1.
-    throw new Error('Unexpected hash algorithm for PQC signature');
+  if (!isCompatibleHashAlgo(signatureAlgo, hashAlgo)) {
+    // The signature hash algo MUST be have digest larger than 256 bits
```
```suggestion
    // The signature hash algo MUST have digest larger than 256 bits
```

twiss commented on src/crypto/public_key/post_quantum/signature/signature.js:
```diff
@@ -52,14 +51,10 @@ export async function verify(signatureAlgo, hashAlgo, eccPublicKey, mldsaPublicK
-      throw new Error('Unsupported signature algorithm');
-  }
+export function isCompatibleHashAlgo(signatureAlgo, hashAlgo) {
+  // The signature hash algo MUST be have digest larger than 256 bits
```
```suggestion
  // The signature hash algo MUST have digest larger than 256 bits
```

twiss reviewed (approved):
The overall logic looks correct to me :+1:
Some nitpicks below. IMHO we don't necessarily have to address them here, but it would be good to refactor a bit when we want to merge it upstream, at least.

larabr committed:
Internal: move signature hash checks to common signature crypto module

To easily reuse the eddsa helpers in the context of MLDSA,
which has different hash requirements.

These signature hash requirements are OpenPGP defined, not
specific to the low-level algorithms.
